File: 2202-sum-of-k-mirror-numbers/sum-of-k-mirror-numbers.java
class Solution {

    // Tạo số đối xứng (palindrome) trong base-10 từ một nửa số `num`
    // Nếu `odd` là true → tạo số có độ dài lẻ (bỏ chữ số giữa)
    long createPalindrome(long num, boolean odd) {
        long x = num;
        if (odd)
            x /= 10;
        while (x > 0) {
            num = num * 10 + x % 10;
            x /= 10;
        }
        return num;
    }

    // Kiểm tra xem số `num` có là palindrome trong base-k hay không
    boolean isPalindrome(long num, int base) {
        StringBuilder sb = new StringBuilder();
        while (num > 0) {
            sb.append((char) (num % base + '0'));
            num /= base;
        }
        String s = sb.toString();
        int i = 0, j = s.length() - 1;
        while (i < j)
            if (s.charAt(i++) != s.charAt(j--))
                return false;
        return true;
    }

    // Hàm chính để tính tổng n số k-mirror nhỏ nhất
    public long kMirror(int k, int n) {
        long sum = 0;
        // len dùng để tạo số đối xứng từ các nửa số có độ dài tăng dần
        for (long len = 1; n > 0; len *= 10) {
            // Tạo số palindrome có độ dài lẻ
            for (long i = len; n > 0 && i < len * 10; i++) {
                long p = createPalindrome(i, true);
                if (isPalindrome(p, k)) {
                    sum += p;
                    n--;
                }
            }
            // Tạo số palindrome có độ dài chẵn
            for (long i = len; n > 0 && i < len * 10; i++) {
                long p = createPalindrome(i, false);
                if (isPalindrome(p, k)) {
                    sum += p;
                    n--;
                }
            }
        }
        return sum;
    }
}

File: 2202-sum-of-k-mirror-numbers/sum-of-k-mirror-numbers.go
func kMirror(k int, n int) int64 {
	var sum int64 = 0
	for length := int64(1); n > 0; length *= 10 {
		// Tạo các số palindrome lẻ
		for i := length; n > 0 && i < length*10; i++ {
			p := createPalindrome(i, true)
			if isPalindrome(p, k) {
				sum += p
				n--
			}
		}
		// Tạo các số palindrome chẵn
		for i := length; n > 0 && i < length*10; i++ {
			p := createPalindrome(i, false)
			if isPalindrome(p, k) {
				sum += p
				n--
			}
		}
	}
	return sum
}

func createPalindrome(num int64, odd bool) int64 {
	x := num
	if odd {
		x /= 10
	}
	for x > 0 {
		num = num*10 + x%10
		x /= 10
	}
	return num
}

func isPalindrome(num int64, base int) bool {
	digits := []int{}
	for num > 0 {
		digits = append(digits, int(num%int64(base)))
		num /= int64(base)
	}
	for i, j := 0, len(digits)-1; i < j; i, j = i+1, j-1 {
		if digits[i] != digits[j] {
			return false
		}
	}
	return true
}
